PKG-473 [8.0]: Implement ccache to reduce compilation time by 50-80%

nogueiraanderson opened the pull request:
## PKG-473: Implement ccache for Jenkins builds to reduce compilation time

This PR adds ccache support to the Percona Server build pipeline, achieving **50-80% reduction in compilation time** by caching compilation artifacts in S3.

## Key Features

### Shared Library Integration
- Integrated with jenkins-pipelines shared library ([PR #3356](https://github.com/Percona-Lab/jenkins-pipelines/pull/3356))
- Uses `ccacheDownload` and `ccacheUpload` functions for S3 cache management
- Clean separation of concerns between build scripts and cache logic

### Smart Cache Key Generation
- **MySQL Version**: Extracted from sources (e.g., 8.0.42, 8.4.5)
- **Toolset**: Compiler + version (e.g., gcc-13.3.0, clang-18.1.8)
- **Build Configuration**: Including special builds (Debug, RelWithDebInfo, ASAN, Valgrind)
- **Build Parameters Type**: Distinguishes between standard/asan/valgrind/ubsan/msan builds

### Configuration Options
- **COMPILER**: Select specific compiler or use system default
- **CACHE_RETENTION_DAYS**: Configure S3 object lifecycle (7/14/21/30 days)
- **CCACHE_MAXSIZE**: Override default 8GB cache size limit
- Environment variables for compression, CPP2 mode, etc.

### Implementation Details
- **Jenkins Pipeline**: 
  - Added COMPILER parameter for toolset selection
  - Extract compiler version and create TOOLSET variable
  - Determine BUILD_PARAMS_TYPE based on ANALYZER_OPTS
- **Docker Build**:
  - Configure ccache environment variables
  - Mount .ccache directory for local cache
  - Use compression level 1 for optimal performance
- **Build Scripts**:
  - Only set CMake launcher options (no local ccache config)
  - Show cache statistics before/after builds

## Performance Results
- **First build**: Populates cache (no time savings)
- **Subsequent builds**: 50-80% faster compilation
- **Cache hit rate**: ~50% in testing (builds ~10,000 files)

## Credits
Implementation incorporates feedback from:
- @inikep (compiler detection, Azure patterns)
- @EvgeniyPatlan (shared library suggestion, shell fixes)  
- @vorsel (shared library review)
- @percona-ysorokin (cache key components, size recommendations)

## Related
- Jira: [PKG-473](https://perconadev.atlassian.net/browse/PKG-473)
- Shared Library PR: [jenkins-pipelines #3356](https://github.com/Percona-Lab/jenkins-pipelines/pull/3356)

EvgeniyPatlan commented on jenkins/pipeline-parallel-mtr.groovy:
```diff
@@ -199,6 +199,80 @@ void checkoutSources() {
     """
 }
 
+def downloadCcache() {
```
maybe it is better to add this function as a shared library so it can be reused for other products?

vorsel replied:
I was about to ask the same question)

nogueiraanderson replied:
Noted. I'll implement it as a shared library.

EvgeniyPatlan commented on jenkins/pipeline-parallel-mtr.groovy:
```diff
@@ -199,6 +199,80 @@ void checkoutSources() {
+    }
+}
+
+def uploadCcache() {
```
maybe it is better to add this function as a shared library so it can be reused for other products?

nogueiraanderson replied:
Done, @EvgeniyPatlan . See associated PR: https://github.com/Percona-Lab/jenkins-pipelines/pull/3356

EvgeniyPatlan commented on local/build-binary:
```diff
@@ -196,6 +227,12 @@ fi
 rm -rf ${INSTALL_DIR}
 
+# Show final ccache statistics
+if [[ "${USE_CCACHE}" == "1" ]] && command -v ccache &> /dev/null; then
```
```suggestion
if [[ "${USE_CCACHE}" == "1" ]] && [[ command -v ccache &> /dev/null ]]; then
```

EvgeniyPatlan commented on local/build-binary:
```diff
@@ -36,6 +36,37 @@ mkdir -p ${BUILD_DIR} ${INSTALL_DIR} ${DOWNLOAD_DIR}
+    echo "Setting up ccache..."
+
+    # Configure ccache (already installed in Docker image)
+    if command -v ccache &> /dev/null; then
```
```suggestion
    if [[ command -v ccache &> /dev/null ]]; then
```

inikep commented on jenkins/pipeline-parallel-mtr.groovy:
```diff
@@ -611,7 +617,34 @@ pipeline {
+                            branch: env.BRANCH,
+                            dockerOs: env.DOCKER_OS,
+                            cmakeBuildType: env.CMAKE_BUILD_TYPE,
+                            compiler: env.COMPILER,
```
The `COMPILER` is undefined by default in Jenkins.
To get best cache hits we need a full compiler version (e.g. `clang-18.1.8` or `gcc-13.3.0`)
In Azure Pipelines we use: https://github.com/percona/percona-server/blob/8.0/azure-pipelines.yml#L363-L367

inikep replied:
Maybe with Jenkins it's not a big deal because we will change a compiler e.g. when docker images will be updated. With Azure we test many compilers with the same docker image.

percona-ysorokin replied:
Completely agree with @inikep here. Object names under which we upload ccache tarballs on AWS S3 should include at least the following 3 components:
1. \<mysql_version> (for instance, `8.0.42` or `8.4.5`)
2. \<toolset> (for instance, `clang-18.1.8` or `gcc-13.3.0`)
3. \<build_configuration> (`Debug` or `RelWithDebInfo`). Ideally ASan (`Debug` + `-DWITH_ASAN=ON`) and Valgrind (`Debug` + `-DWITH_Valgrind=ON`) should also have unique \<build_configuration> labels (e.g. `ASasn` and `Valgrind`).

nogueiraanderson replied:
Done. Thanks for the feedback @inikep and @percona-ysorokin!

inikep commented on jenkins/pipeline-parallel-mtr.groovy:
```diff
@@ -611,7 +617,34 @@ pipeline {
+                            awsCredentialsId: AWS_CREDENTIALS_ID,
+                            s3Bucket: S3_ROOT_DIR + '/',
+                            workspace: env.WORKSPACE,
+                            branch: env.BRANCH,
```
The `PARENT_BRANCH` in Azure is misleading because it basically should be a server version (`5.7`, `8.0`, `8.4`). The idea behind it is that we want to share cache for the same server version even when a branch name is different.
Even better solution would be to distinguish `8.4.4`, `8.4.5` and so on:
We have this information in `MYSQL_VERSION` variable in https://github.com/Percona-Lab/ps-build/blob/8.0/local/build-binary#L126

inikep commented:
What about cache retention? Shouldn't we remove caches when they are not used/changed in 1 month?

percona-ysorokin commented on docker/run-build:
```diff
@@ -33,6 +33,14 @@ docker run --rm \
+    -e CCACHE_COMPRESS=1 \
+    -e CCACHE_COMPRESSLEVEL=9 \
+    -e CCACHE_CPP2=1 \
+    -e CCACHE_MAXSIZE=2G \
```
1. Not sure is 2GB is enough for our builds.
2. May be it is better to change persistent ccache options via command line (`ccache --max-size=8G`, for instance)

percona-ysorokin commented on local/build-binary:
```diff
@@ -36,6 +36,37 @@ mkdir -p ${BUILD_DIR} ${INSTALL_DIR} ${DOWNLOAD_DIR}
+        ccache --version
+        ccache --max-size="${CCACHE_MAXSIZE:-2G}"
+        ccache --set-config compression=true
+        ccache --set-config compression_level=9
```
Probably, no need to set up environment variables `CCACHE_COMPRESS=1` and `CCACHE_COMPRESSLEVEL=9` if you set these values here.

percona-ysorokin commented on local/build-binary:
```diff
@@ -36,6 +36,37 @@ mkdir -p ${BUILD_DIR} ${INSTALL_DIR} ${DOWNLOAD_DIR}
+    # Configure ccache (already installed in Docker image)
+    if command -v ccache &> /dev/null; then
+        ccache --version
+        ccache --max-size="${CCACHE_MAXSIZE:-2G}"
```
The same question regarding the 2GB size here.
For instance, the total size of all `.o` files in my clang-19 ASan build dir is 3.7 GB.
For gcc-14 ASan - 8GB.
May be with compression it will be smaller, but to be safe I would put 8GB limit there

nogueiraanderson replied:
Thanks for the input, @percona-ysorokin .

percona-ysorokin commented:
> What about cache retention? Shouldn't we remove caches when they are not used/changed in 1 month?

I think it is possible to set this on S3 bucket / object level https://docs.aws.amazon.com/AmazonS3/latest/userguide/how-to-set-lifecycle-configuration-intro.html

inikep commented:
@nogueiraanderson My recent PR at https://github.com/Percona-Lab/ps-build/pull/465/files#diff-cbe31d84b2f944ad01d8f4adc62f6dc7303d467790017282389f014e4453734cR10 introduces `SERVER_VERSION` variable which you may use for `mysqlVersion`. I used `SERVER_VERSION` because it's also a version of Percona Server for MySQL.

nogueiraanderson commented:
## All Review Comments Addressed

All feedback has been addressed across multiple commits:

**@EvgeniyPatlan:**
- Shared library implementation: Created in jenkins-pipelines PR #3356
- Shell syntax: Fixed in commit c406c4c - now using double brackets with command substitution

**@inikep:**
- Compiler auto-detection: Added automatic detection when COMPILER is undefined (commit efb9767)
- Full compiler version: Using Azure Pipelines tested approach (commit ceb7732)
- SERVER_VERSION: Updated to use SERVER_VERSION instead of MYSQL_VERSION (commit db59262)
- Cache retention: Implemented via S3 lifecycle rules (PKG-769 completed)
- COMPILER parameter: Restored to Jenkins job definition (commit fc1a0ee)

**@percona-ysorokin:**
- Cache size: Increased to 8GB as suggested
- Build type suffixes: Added -asan/-valgrind suffixes for unique cache keys
- Environment variables: Removed redundant CCACHE_COMPRESS variables
- S3 lifecycle: Implemented tag-based retention (7, 14, 21, 30 days)

All commits are now visible in the PR.

**Related Jira:**
- [PKG-473](https://perconadev.atlassian.net/browse/PKG-473)
- [PKG-769](https://perconadev.atlassian.net/browse/PKG-769)

inikep commented on local/build-binary:
```diff
@@ -36,6 +36,37 @@ mkdir -p ${BUILD_DIR} ${INSTALL_DIR} ${DOWNLOAD_DIR}
+    # Configure ccache (already installed in Docker image)
+    if [[ -n $(command -v ccache) ]]; then
+        ccache --version
+        ccache --max-size=8G
```
We can't change `ccache` settings in `local/build-binary` because when somebody will run this script locally without Docker it will change his/her `ccache`. I have `max_size = 256.0 GB` for example.

nogueiraanderson committed:
PKG-473: Implement ccache for Jenkins builds to reduce compilation time

This implementation adds ccache support to significantly reduce build times (50-80% improvement)
by caching compilation artifacts in S3.

Key features:
- Integrated with jenkins-pipelines shared library (ccacheDownload/ccacheUpload functions)
- Smart cache key generation using mysql version, toolset, and build configuration
- Support for different build types (Debug, RelWithDebInfo, ASAN, Valgrind, etc.)
- Configurable cache parameters through Jenkins UI
- Automatic compiler detection and version extraction
- S3-based cache storage with lifecycle management

Changes:
- Add COMPILER parameter to pipeline YAML files for toolset selection
- Extract compiler version and create TOOLSET variable (e.g., gcc-13.3.0, clang-18.1.8)
- Add BUILD_PARAMS_TYPE based on ANALYZER_OPTS for specialized builds
- Add Jenkins parameters: CACHE_RETENTION_DAYS and CCACHE_MAXSIZE
- Configure ccache environment variables in docker/run-build
- Update build-binary to use ccache with CMake
- Remove redundant local ccache configuration
- Organize ccache function parameters alphabetically for consistency

The implementation follows Azure Pipelines patterns and incorporates feedback from
@inikep, @EvgeniyPatlan, @vorsel, and @percona-ysorokin.

Co-developed with jenkins-pipelines PR #3356 for shared library functions.

nogueiraanderson committed:
PKG-473: Remove user-configurable cache parameters

- Remove CACHE_RETENTION_DAYS and CCACHE_MAXSIZE Jenkins parameters
- Add constants at top of pipeline: CCACHE_MAXSIZE=8G, retention 60/120 days
- Hardcode CCACHE_MAXSIZE=8G in docker/run-build script
- Update ccache function calls to remove deprecated parameters
- Simplify configuration based on reviewer feedback from Przemyslaw and Yura

nogueiraanderson committed:
PKG-473: Add documentation for cache configuration constants

- Document that CCACHE_MAXSIZE of 8GB is sufficient for all builds
- Clarify retention periods: 60 days normal, 120 days sanitizers
- Add NOTE about S3 lifecycle rule configuration requirement

nogueiraanderson committed:
PKG-473: Address additional review feedback from Przemyslaw

- Change TOOLSET creation from shell script to Groovy code
- Remove COMPILER parameter (always use system default)
- Change USE_CCACHE from 'true' to 'yes' for consistency with other params
- Remove duplicated BUILD_TYPE_SUFFIX logic (BUILD_PARAMS_TYPE is sufficient)
- Update shared library to use 'yes' instead of 'true' for USE_CCACHE

nogueiraanderson committed:
PKG-473: Fix USE_CCACHE check in build-binary for consistency

Changed from 'true' to 'yes' to match all other USE_CCACHE checks

nogueiraanderson committed:
PKG-473: Fix remaining USE_CCACHE check for consistency

Changed the final ccache statistics check from 'true' to 'yes' to match all other USE_CCACHE checks throughout the codebase

nogueiraanderson committed:
PKG-473: Pass cloud parameter to ccache functions for Hetzner support

- Pass params.CLOUD to ccacheDownload and ccacheUpload functions
- Enables automatic switching between AWS and Hetzner Object Storage
- Works with updated shared library functions from jenkins-pipelines

inikep commented on jenkins/param-parallel-mtr.yml:
```diff
@@ -47,11 +47,6 @@
-        name: COMPILER
-        choices:
-        - default
-        description: compiler version e.g. gcc-15 or clang-20
```
We shouldn't remove it (and keep `COMPILER`).

inikep commented on jenkins/pipeline-parallel-mtr.groovy:
```diff
@@ -606,8 +621,97 @@ pipeline {
+                                def mysqlVersion = sh(returnStdout: true, script: """
+                                    echo "${versionFile}" | grep -E '^MYSQL_VERSION_MAJOR|^MYSQL_VERSION_MINOR|^MYSQL_VERSION_PATCH' | cut -d= -f2 | paste -sd'.' -
+                                """).trim()
+                                env.MYSQL_VERSION = mysqlVersion
```
All of it is already done with:
```
SERVER_VERSION = validatePsBranch()
```
at https://github.com/Percona-Lab/ps-build/blob/8.0/jenkins/pipeline-parallel-mtr.groovy#L585C28-L585C56
I'm using `SERVER_VERSION` because it's either version of MySQL or Percona Server for MySQL.

inikep commented on jenkins/pipeline-parallel-mtr.yml:
```diff
@@ -43,6 +43,11 @@
+        name: COMPILER
+        choices:
+        - default
+        description: compiler version (default = use system compiler)
```
COMPILER is already there (look several lines below).

nogueiraanderson committed:
PKG-473: Address inikep's review feedback

- Restore COMPILER parameter that was accidentally removed
- Remove duplicate MySQL version extraction code
- Use existing SERVER_VERSION from validatePsBranch() instead

As noted by @inikep:
- COMPILER parameter is required and should not be removed
- SERVER_VERSION already extracts the version information we need

nogueiraanderson committed:
PKG-473: Remove duplicate COMPILER parameter in pipeline YAML

As noted by @inikep, COMPILER parameter already exists in the file
(around line 63). Removed the duplicate entry that was added earlier.

nogueiraanderson committed:
PKG-473: Improve S3 cache retention documentation

- Add detailed explanation of S3 lifecycle configuration requirements
- Clarify that tags are automatically set by ccacheUpload()
- Reference PKG-769 for S3 bucket configuration details
- Explain tag-based expiration rules for 60 and 120 day retention

nogueiraanderson committed:
PKG-473: Add Hetzner Object Storage lifecycle configuration notes

- Document that Hetzner doesn't support tag-based lifecycle rules
- Suggest prefix-based organization for Hetzner deployments
- Note versioning requirements for lifecycle policies
- Provide alternative cache organization strategy for Hetzner S3

nogueiraanderson committed:
PKG-473: Clarify Hetzner lifecycle limitations are not addressed

- Remove misleading suggestion about prefix-based organization
- Add TODO note that current tag-based implementation won't work on Hetzner
- Clarify that prefix-based paths would need to be implemented

nogueiraanderson committed:
PKG-473: Clarify TODO for Hetzner expiry date implementation

- Specify that prefix-based paths are needed for Hetzner lifecycle policies
- Emphasize that this enables automatic object expiration
- Make it clear this is required for Hetzner deployments to have cache cleanup

nogueiraanderson committed:
PKG-473: Remove Dockerfile.jjb-deploy from PR

This file was accidentally included and is not part of the ccache implementation

inikep commented on docker/run-build:
```diff
@@ -52,6 +53,15 @@ docker run --rm \
+    export CCACHE_COMPRESS='${CCACHE_COMPRESS:-1}'
+    export CCACHE_COMPRESSLEVEL='${CCACHE_COMPRESSLEVEL:-1}'
+    export CCACHE_CPP2='${CCACHE_CPP2:-1}'
+    export CCACHE_MAXSIZE='8G'
```
This should depend on `CCACHE_MAXSIZE` from groovy file.

inikep commented on jenkins/param-parallel-mtr.yml:
```diff
@@ -296,19 +294,3 @@
       - percona-server-{ps_version}-param-parallel-mtr
-
-- project:
-    name: param-parallel-mtr-9.x
```
This reverts `jenkins/param-parallel-mtr.yml` to some state in the past or 8.x branch. 
Probably you shouldn't touch this file at all.

nogueiraanderson committed:
fix: revert changes to wrong file

nogueiraanderson committed:
Fix CCACHE_MAXSIZE to depend on groovy file constant

- Pass CCACHE_MAXSIZE from groovy constant to docker build script
- Update docker/run-build to use passed value with fallback to 8G
- Addresses review comment from @inikep

inikep reviewed (approved):
LGTM

nogueiraanderson committed:
PKG-473: Pass explicit cache retention days to ccacheUpload

- Add logic to determine retention based on BUILD_PARAMS_TYPE
- Pass 120 days for ASAN/Valgrind builds, 60 days for normal builds
- Update comments to clarify retention policy ownership
- Pipeline now controls business logic while shared library provides utility

percona-ysorokin reviewed (approved):
LGTM
